bt-core/src/main/java/bt/torrent/messaging/TorrentWorker.java

```diff
@@ -126,18 +135,38 @@ private BitfieldBasedStatistics getStatistics() {
      * @since 1.0
      */
     private void addPeer(ConnectionKey connectionKey) {
-        PieceAnnouncingPeerWorker worker = createPeerWorker(connectionKey);
-        PieceAnnouncingPeerWorker existing = peerMap.putIfAbsent(connectionKey, worker);
-        if (existing == null) {
-            dispatcher.addMessageConsumer(connectionKey, message -> consume(connectionKey, message));
-            dispatcher.addMessageSupplier(connectionKey, () -> produce(connectionKey));
+        if (tryAddPeerWithLimits()) {
+            PieceAnnouncingPeerWorker worker = Objects.requireNonNull(createPeerWorker(connectionKey));
```

Comment:
This requireNonNull is necessary to ensure that the count does not get off. If we put null into the Map, it will increase the count by 1. If afterwards, it is worker is non-null, we'll replace it, increasing the count again, double counting it.

I don't believe that worker ever can be null, but I think adding a null check with a comment here is smart. Alternatively, we could correctly handle the case where it is null, and decrement the count `if(existing != null || worker == null)`. This is a bit clunky because count will differ from `peerMap.size()` so I prefer having the assertion. What do you think?

Reply:
Nevermind - ConcurrentHashMap cannot hold null values.